### ISSUE TITLE
os.rename throws exception when renaming directories on lower overlayfs layers

### ISSUE THREAD

**Issue by quasiben** [conda/conda#6711]
OS: CentOS Linux release 7.2.1511 (Core)
docker version: 1.12.6
docker storage driver: overlayfs

Docker Image has existing conda environment baked into layer

```
    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/exceptions.py", line 640, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/main.py", line 140, in _main
        exit_code = args.func(args, p)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/main_install.py", line 80, in execute
        install(args, parser, 'install')
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/install.py", line 326, in install
        execute_actions(actions, index, verbose=not context.quiet)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/plan.py", line 830, in execute_actions
        execute_instructions(plan, index, verbose)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/instructions.py", line 247, in execute_instructions
        cmd(state, arg)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/instructions.py", line 108, in UNLINKLINKTRANSACTION_CMD
        txn.execute()
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/link.py", line 297, in execute
        rollback_excs,
    conda.CondaMultiError: [Errno 22] Invalid argument: '/opt/continuum/anaconda/envs/anaconda44_py35/lib/python3.5/version.txt' -> '/opt/continuum/anaconda/envs/anaconda44_py35/lib/python3.5/version.txt.c~'
```

cc @msarahan @kalefranz @mmarchetti @dsludwig

**Comment by kalefranz** [comment]
Can you give me the full stack trace?  That just looks like part of it...

> `During handling of the above exception, another exception occurred:`

**Comment by quasiben** [comment]
```
$ /opt/continuum/anaconda/envs/anaconda44_py35/bin/conda install --yes --prefix /opt/continuum/anaconda/envs/anaconda44_py35 --channel anaconda-enterprise anaconda-enterprise jupyter_kernel_gateway anaconda-enterprise-web-publisher --debug`




    Traceback (most recent call last):
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/link.py", line 326, in _execute_actions
        action.execute()
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/path_actions.py", line 620, in execute
        backoff_rename(self.target_full_path, self.holding_full_path, force=True)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/gateways/disk/update.py", line 53, in backoff_rename
        exp_backoff_fn(rename, source_path, destination_path, force)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/gateways/disk/__init__.py", line 21, in exp_backoff_fn
        return fn(*args, **kwargs)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/gateways/disk/update.py", line 47, in rename
        os_rename(source_path, destination_path)
    OSError: [Errno 22] Invalid argument: '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project' -> '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project.c~'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/link.py", line 281, in execute
        pkg_data, actions)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/link.py", line 344, in _execute_actions
        reverse_excs,
    conda.CondaMultiError: [Errno 22] Invalid argument: '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project' -> '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project.c~'


    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/exceptions.py", line 640, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/main.py", line 140, in _main
        exit_code = args.func(args, p)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/main_install.py", line 80, in execute
        install(args, parser, 'install')
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/cli/install.py", line 326, in install
        execute_actions(actions, index, verbose=not context.quiet)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/plan.py", line 830, in execute_actions
        execute_instructions(plan, index, verbose)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/instructions.py", line 247, in execute_instructions
        cmd(state, arg)
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/instructions.py", line 108, in UNLINKLINKTRANSACTION_CMD
        txn.execute()
      File "/opt/continuum/anaconda/lib/python3.6/site-packages/conda/core/link.py", line 297, in execute
        rollback_excs,
    conda.CondaMultiError: [Errno 22] Invalid argument: '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project' -> '/opt/continuum/anaconda/envs/anaconda44_py35/bin/anaconda-project.c~'
```

**Comment by kalefranz** [comment]
Related reference: https://github.com/moby/moby/issues/25409

**Comment by kalefranz** [comment]
errno 22 is `EINVAL`

**Comment by kalefranz** [comment]
#define EINVAL          22      /* Invalid argument */

**Comment by kalefranz** [comment]
NPM discussion: https://github.com/npm/npm/issues/9863

**Comment by kalefranz** [comment]
Might no just be errno 22.  moby/moby#25409 has

    OSError: [Errno 18] Invalid cross-device link: '/foo' -> '/foo2'

which is `EXDEV`.

**Comment by quasiben** [comment]
fixed in PR #6712
